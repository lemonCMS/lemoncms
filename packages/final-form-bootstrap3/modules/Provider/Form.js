import React from "react";
import PropTypes from "prop-types";
import { Form as FinalForm } from "react-final-form";
import arrayMutators from "final-form-arrays";
import ContextWrapper from "../Context/ContextWrapper";

const Form = props => {
  const { children, className, debug, readOnly, listen, ...formProps } = props;

  return (
    <FinalForm
      {...formProps}
      mutators={{
        ...arrayMutators
      }}
      render={({ handleSubmit, ...rest }) => {
        return (
          <ContextWrapper
            {...rest}
            debug={debug}
            listen={listen}
            readOnly={readOnly}
          >
            <form onSubmit={handleSubmit} className={className}>
              {children}
            </form>
          </ContextWrapper>
        );
      }}
    />
  );
};

Form.propTypes = {
  className: PropTypes.string,
  validate: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  debug: PropTypes.bool,
  readOnly: PropTypes.bool,
  listen: PropTypes.func,
  initialValues: PropTypes.objectOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.array,
      PropTypes.object,
      PropTypes.bool
    ])
  )
};

Form.defaultProps = {
  className: "",
  validate: () => {},
  listen: () => {},
  debug: false,
  readOnly: false,
  initialValues: {}
};

export default Form;
