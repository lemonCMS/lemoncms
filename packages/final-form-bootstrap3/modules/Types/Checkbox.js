import React, { Component } from "react";
import ReactDOMServer from "react-dom/server";
import PropTypes from "prop-types";
import memoize from "memoize-one";
import CheckboxAlias from "react-bootstrap/lib/Checkbox";
import Col from "react-bootstrap/lib/Col";
import Row from "react-bootstrap/lib/Row";
import context from "../decorators/context";
import fieldGroup from "./fieldGroup";
import Filter from "./components/Filter";

class Checkbox extends Component {
  state = {
    values: [],
    filterText: ""
  };

  componentDidMount() {
    const {
      input: { value }
    } = this.props;

    if (typeof value !== "undefined") {
      this.setState({ values: value });
    }
  }

  emit = (event, checkboxValue, single = false) => {
    const {
      input: { onChange, value }
    } = this.props;

    if (single) {
      // Single checkbox
      if (event.target.checked) {
        // Set the value from the checkbox
        onChange(checkboxValue);
        return;
      }

      // Remove value
      onChange("");
      return;
    }

    const clone = [...value].filter(filter => filter); // make a shallow clone
    // Multiple checkboxes
    if (event.target.checked) {
      // See if item is already in the value array, if so ignore it.
      if (clone.indexOf(checkboxValue) > -1) {
        return;
      }

      // Value does not exist, set it
      clone.push(checkboxValue);
      onChange(clone);
      return true;
    }

    // Remove value
    // If it does not exist, ignore it
    if (clone.indexOf(checkboxValue) === -1) {
      return;
    }

    // the value exists and need te be removed.
    clone.splice(clone.indexOf(checkboxValue), 1);
    onChange(clone);
  };

  filtered = memoize((list, filterText) => {
    return list.filter(item => {
      if (!filterText && item.type !== "option") {
        return true;
      }

      if (typeof item.props.children === "string") {
        return (
          item.props &&
          item.type === "option" &&
          item.props.children.includes(filterText)
        );
      }

      return ReactDOMServer.renderToString(item.props.children).includes(
        filterText
      );
    });
  });

  handleChange = event => {
    this.setState({ filterText: event.target.value });
  };

  clearFilterText = event => {
    this.setState({ filterText: "" });
  };

  createCheckBoxes = children => {
    const { columns, filter, input } = this.props;
    if (
      children &&
      Array.isArray(children) &&
      React.isValidElement(children[0]) &&
      children[0].type === "option"
    ) {
      const list = children.map(option => {
        if (option.type !== "option") {
          return option;
        }
        const { value, children } = option.props;

        return (
          <CheckboxAlias
            key={value}
            checked={input.value.indexOf(value) > -1}
            onChange={event => {
              this.emit(event, value);
            }}
          >
            {children}
          </CheckboxAlias>
        );
      });

      if (columns === 1) {
        return list;
      }

      const itemsPerColumn = Math.ceil(list.length / columns);
      // create new array of length with undefined values
      const cols = Array.apply(null, Array(columns)).map(() => {});
      const display = cols.map((col, index) => {
        const start = index * itemsPerColumn;
        const end = start + itemsPerColumn;
        return (
          <Col md={Math.round(12 / columns)} key={index}>
            {list.slice(start, end)}
          </Col>
        );
      });

      return <Row>{display}</Row>;
    }

    // Filtered returned nothing
    if (filter) {
      return <div>No results</div>;
    }

    return (
      <CheckboxAlias
        checked={input.value}
        onChange={event => {
          this.emit(event, true, true);
        }}
      >
        {children}
      </CheckboxAlias>
    );
  };

  render() {
    const { filter, placeholderFilter, children } = this.props;
    const { filterText } = this.state;
    if (filter) {
      const filteredList = this.filtered(children, this.state.filterText);
      return (
        <>
          <Filter
            filterText={filterText}
            clearFilterText={this.clearFilterText}
            handleChange={this.handleChange}
            placeholderFilter={placeholderFilter}
          />
          {this.createCheckBoxes(filteredList)}
        </>
      );
    }

    return <>{this.createCheckBoxes(children)}</>;
  }
}

Checkbox.propTypes = {
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.array
    ])
  }),
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string,
    PropTypes.object,
    PropTypes.element
  ]),
  placeholderFilter: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  columns: PropTypes.number,
  filter: PropTypes.bool,
  computedInvalid: PropTypes.bool.isRequired
};

Checkbox.defaultProps = {
  input: {},
  children: [],
  label: null,
  help: null,
  columns: 1,
  filter: false,
  placeholderFilter: "Filter"
};

export default context()(fieldGroup(Checkbox));
