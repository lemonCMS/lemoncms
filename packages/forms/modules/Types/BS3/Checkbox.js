import React from "react";
import PropTypes from "prop-types";
import { FormControl } from "react-bootstrap-v3";
import context from "../../decorators/context";
import fieldGroup from "./fieldGroup";

const Checkbox = props => {
  const { input } = props;

  return <FormControl {...input} type={"password"} />;
};

Checkbox.propTypes = {
  input: PropTypes.oneOfType([PropTypes.object]),
  placeholder: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired
};

Checkbox.defaultProps = {
  input: {},
  label: null,
  help: null,
  placeholder: null
};

export default context(fieldGroup(Checkbox));
