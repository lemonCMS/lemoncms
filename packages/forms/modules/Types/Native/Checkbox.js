import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import context from '../../decorators/context';
import field from '../../decorators/field';

const Checkbox = props => {
  const {
    input, isDisabled, multiple, children, formControl, computedInvalid
  } = props;

  const radioList = Array.isArray(children) ? children : [children];

  const onChange = (event, key, option) => {
    if (multiple) {
      const newValue = (input.value instanceof Array ? input.value : [input.value]).filter(Boolean);
      if (event.target.checked) {
        newValue.push(option.props.value);
      } else {
        newValue.splice(newValue.indexOf(option.props.value), 1);
      }
      return input.onChange(newValue);
    }
    return input.onChange(event.target.checked ? option.props.value : false);
  };

  const list = radioList.map((option, key) => {
    const name = multiple ? `${input.name}[${key}]` : input.name;
    const value = (Array.isArray(input.value) ? input.value : [input.value]).filter(val => String(val));
    const checked = value.indexOf(option.props.value) !== -1;

    return (
      <div className="form-check" key={option.props.value}>
        <input
          type="checkbox"
          name={name}
          disabled={isDisabled}
          id={`${input.name}_${key}`}
          value={option.props.value}
          className={classNames('form-check-label')}
          checked={checked}
          onChange={event => onChange(event, key, option)}
        />{' '}
        <label htmlFor={`${input.name}_${key}`} className={classNames('form-check-label')}>
          {option.props.children}
        </label>
      </div>
    );
  });

  return <div className={classNames('ffc-input', { 'is-invalid': computedInvalid }, formControl)}>{list}</div>;
};

Checkbox.propTypes = {
  input: PropTypes.oneOfType([PropTypes.object]),
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  formControl: PropTypes.string,
  multiple: PropTypes.bool,
  isDisabled: PropTypes.bool,
  computedInvalid: PropTypes.bool.isRequired
};

Checkbox.defaultProps = {
  input: {},
  formControl: null,
  multiple: false,
  isDisabled: false
};

export default context(field(Checkbox));
