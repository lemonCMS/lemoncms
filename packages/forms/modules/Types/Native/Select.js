import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import context from '../../decorators/context';
import field from '../../decorators/field';

const Select = props => {
  const {
    input, children, placeholder, formControl, computedInvalid
  } = props;

  const {value, ...rest} = input;


  const options = Array.isArray(children) ? children : [children];
  const list = options.map((option, key) => (
    <option
      key={key}
      value={option.props.value}
      selected={String(input.value) === String(option.props.value)}
    >
      {option.props.children}
    </option>
  ));

  return (
    <select
      placeholder={placeholder}
      className={classNames('ffc-input', { 'is-invalid': computedInvalid }, formControl)}
      {...rest}
    >
      {list}
    </select>
  );
};

Select.propTypes = {
  input: PropTypes.oneOfType([PropTypes.object]),
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  placeholder: PropTypes.string,
  formControl: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired
};

Select.defaultProps = {
  input: {},
  placeholder: null,
  formControl: null
};

export default context(field(Select));
