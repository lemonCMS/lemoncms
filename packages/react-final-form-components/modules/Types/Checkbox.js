import React, { Component } from "react";
import ReactDOMServer from "react-dom/server";
import PropTypes from "prop-types";
import memoize from "memoize-one";
import CheckboxAlias from "react-bootstrap/lib/Checkbox";
import Col from "react-bootstrap/lib/Col";
import Row from "react-bootstrap/lib/Row";
import context from "../decorators/context";
import fieldGroup from "./fieldGroup";
import Filter from "./components/Filter";

class Checkbox extends Component {
  state = {
    values: [],
    filterText: ""
  };

  componentDidMount() {
    const {
      input: { value }
    } = this.props;

    if (typeof value !== "undefined") {
      this.setState({ values: value });
    }
  }

  emit = (event, checkboxValue, single = false) => {
    const {
      input: { onChange, value }
    } = this.props;

    if (single) {
      // Single checkbox
      if (event.target.checked) {
        // Set the value from the checkbox
        onChange(checkboxValue);
        return;
      }

      // Remove value
      onChange("");
      return;
    }

    const clone = [...value].filter(filter => filter); // make a shallow clone
    // Multiple checkboxes
    if (event.target.checked) {
      // See if item is already in the value array, if so ignore it.
      if (clone.indexOf(checkboxValue) > -1) {
        return;
      }

      // Value does not exist, set it
      clone.push(checkboxValue);
      onChange(clone);
      return true;
    }

    // Remove value
    // If it does not exist, ignore it
    if (clone.indexOf(checkboxValue) === -1) {
      return;
    }

    // the value exists and need te be removed.
    clone.splice(clone.indexOf(checkboxValue), 1);
    onChange(clone);
  };

  filtered = memoize((list, filterText) => {
    return list.filter(item => {
      if (!filterText && item.type !== "option") {
        return true;
      }

      if (typeof item.props.children === "string") {
        return (
          item.props &&
          item.type === "option" &&
          item.props.children.includes(filterText)
        );
      }

      return ReactDOMServer.renderToString(item.props.children).includes(
        filterText
      );
    });
  });

  handleChange = event => {
    this.setState({ filterText: event.target.value });
  };

  clearFilterText = event => {
    this.setState({ filterText: "" });
  };

  createCheckBoxes = children => {
    const { columns, filter, input, isDisabled } = this.props;
    if (
      children &&
      Array.isArray(children) &&
      React.isValidElement(children[0]) &&
      children[0].type === "option"
    ) {
      const list = children.map(option => {
        if (option.type !== "option") {
          return option;
        }
        const { value, children } = option.props;

        return (
          <CheckboxAlias
            key={value}
            disabled={isDisabled}
            checked={input.value.indexOf(value) > -1}
            onChange={event => {
              this.emit(event, value);
            }}
          >
            {children}
          </CheckboxAlias>
        );
      });

      if (columns === 1) {
        return list;
      }

      const itemsPerColumn = Math.ceil(list.length / columns);
      // create new array of length with undefined values
      const cols = Array.apply(null, Array(columns)).map(() => {});
      const display = cols.map((col, index) => {
        const start = index * itemsPerColumn;
        const end = start + itemsPerColumn;
        return (
          <Col sm={Math.round(12 / columns)} key={index}>
            {list.slice(start, end)}
          </Col>
        );
      });

      return <Row>{display}</Row>;
    }

    // Filtered returned nothing
    if (filter) {
      return <div>No results</div>;
    }

    return (
      <CheckboxAlias
        checked={input.value}
        disabled={isDisabled}
        onChange={event => {
          this.emit(event, true, true);
        }}
      >
        {children}
      </CheckboxAlias>
    );
  };

  render() {
    const { filter, placeholderFilter, children, isDisabled } = this.props;
    const { filterText } = this.state;
    if (filter) {
      const filteredList = this.filtered(children, this.state.filterText);
      return (
        <>
          <Filter
            isDisabled={isDisabled}
            filterText={filterText}
            clearFilterText={this.clearFilterText}
            handleChange={this.handleChange}
            placeholderFilter={placeholderFilter}
          />
          {this.createCheckBoxes(filteredList)}
        </>
      );
    }

    return <>{this.createCheckBoxes(children)}</>;
  }
}

Checkbox.propTypes = {
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.array
    ])
  }),
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string,
    PropTypes.object,
    PropTypes.element
  ]),
  disabled: PropTypes.func,
  isDisabled: PropTypes.bool,
  placeholderFilter: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  mandatory: PropTypes.bool,
  columns: PropTypes.number,
  filter: PropTypes.bool,
  computedInvalid: PropTypes.bool.isRequired,
  layout: PropTypes.shape({
    label: PropTypes.object,
    field: PropTypes.object
  })
};

Checkbox.defaultProps = {
  input: {},
  children: [],
  label: null,
  help: null,
  columns: 1,
  filter: false,
  placeholderFilter: "Filter",
  disabled: null,
  isDisabled: false,
  mandatory: false,
  layout: null
};

export default context()(fieldGroup(Checkbox));
