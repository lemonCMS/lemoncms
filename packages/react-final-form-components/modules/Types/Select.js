import React from "react";
import PropTypes from "prop-types";
import { FormControl } from "react-bootstrap";
import context from "../decorators/context";
import fieldGroup from "./fieldGroup";

const Select = props => {
  const { input, children, isDisabled, multiple } = props;
  return (
    <FormControl
      {...input}
      componentClass="select"
      disabled={isDisabled}
      multiple={multiple}
    >
      {children}
    </FormControl>
  );
};

Select.propTypes = {
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.array
    ])
  }),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]),
  disabled: PropTypes.func,
  isDisabled: PropTypes.bool,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  help: PropTypes.string,
  computedInvalid: PropTypes.bool.isRequired,
  multiple: PropTypes.bool,
  layout: PropTypes.shape({
    label: PropTypes.object,
    field: PropTypes.object
  })
};

Select.defaultProps = {
  input: {},
  label: null,
  help: null,
  placeholder: null,
  multiple: false,
  disabled: null,
  isDisabled: false,
  layout: null
};

export default context({ type: "select" })(fieldGroup(Select));
