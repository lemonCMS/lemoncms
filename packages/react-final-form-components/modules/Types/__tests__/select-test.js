import renderer from "react-test-renderer";
import { mount } from "enzyme";
import Form from "../../Provider/Form";
import Select from "../Select";
import React from "react";
import Checkbox from "../Checkbox";

describe("BS3 selectbox", () => {
  it("should render snapshot", () => {
    const tree = renderer.create(
      <Form onSubmit={() => {}}>
        <Select name={"field"} label={"Kids"}>
          <option value={"hailey"}>Hailey</option>
          <option value={"aiden"}>Aiden</option>
          <option value={"mayson"}>Mayson</option>
        </Select>
      </Form>
    );
    const output = tree.toJSON();
    expect(output).toMatchSnapshot();
  });

  it("should emit single value", () => {
    const submit = jest.fn();
    const component = mount(
      <Form onSubmit={submit}>
        <Select name={"field"} label={"Kids"}>
          <option value={"hailey"}>Hailey</option>
          <option value={"aiden"}>Aiden</option>
          <option value={"mayson"}>Mayson</option>
        </Select>
      </Form>
    );

    const option = component.find("option");
    expect(option).toHaveLength(3);
    expect(option.at(0).props()).toHaveProperty("value", "hailey");
    expect(option.at(1).props()).toHaveProperty("value", "aiden");
    expect(option.at(2).props()).toHaveProperty("value", "mayson");

    component
      .find("select")
      .simulate("change", { target: { name: "field", value: "aiden" } });
    component.find("form").simulate("submit");

    expect(submit).toHaveBeenCalledTimes(1);
    expect(submit).toHaveBeenCalledWith(
      { field: "aiden" },
      expect.any(Object),
      expect.any(Function)
    );
  });

  it("should emit multiple values", () => {
    const submit = jest.fn();
    const component = mount(
      <Form onSubmit={submit} initialValues={{ field: [] }}>
        <Select name={"field"} label={"Kids"} multiple>
          <option value={"hailey"}>Hailey</option>
          <option value={"aiden"}>Aiden</option>
          <option value={"mayson"}>Mayson</option>
        </Select>
      </Form>
    );

    const option = component.find("option");
    expect(option).toHaveLength(3);
    expect(option.at(0).props()).toHaveProperty("value", "hailey");
    expect(option.at(1).props()).toHaveProperty("value", "aiden");
    expect(option.at(2).props()).toHaveProperty("value", "mayson");

    component.find("select").simulate("change", {
      target: { name: "field", value: ["hailey", "mayson"] }
    });
    component.find("form").simulate("submit");

    expect(submit).toHaveBeenCalledTimes(1);
    expect(submit).toHaveBeenCalledWith(
      { field: ["hailey", "mayson"] },
      expect.any(Object),
      expect.any(Function)
    );
  });
});
