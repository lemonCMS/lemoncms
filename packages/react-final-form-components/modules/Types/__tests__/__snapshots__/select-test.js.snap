// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BS3 selectbox should render snapshot 1`] = `
<form
  className=""
  onSubmit={[Function]}
>
  <div
    className="form-group has-feedback"
  >
    <label
      className="col-sm-4 control-label"
    >
      Kids
    </label>
    <div
      className="col-sm-8"
    >
      <select
        className="form-control"
        disabled={null}
        multiple={false}
        name="field"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value=""
      >
        <option
          value="hailey"
        >
          Hailey
        </option>
        <option
          value="aiden"
        >
          Aiden
        </option>
        <option
          value="mayson"
        >
          Mayson
        </option>
      </select>
    </div>
    <div
      className="clearfix"
    />
  </div>
</form>
`;
